### PR TITLE
Testing

### PR DESCRIPTION
**Pull request by zineanteoh**
This PR consists of setting up GitHub Actions CI for Unit Testing (Jest) and E2E testing (Detox). 

## TL;DR

This was a pain to set up. So many errors, so little reliable documentations. 

## Why it's a pain: 
To do component testing, the official React Native website [recommends](https://reactnative.dev/docs/testing-overview#component-tests) using either: 
1. React's [Test Renderer](https://legacy.reactjs.org/docs/test-renderer.html) (which has no documentation on how to set up), 
6. or [React Native Testing Library](https://callstack.github.io/react-native-testing-library/) (which uses React's Test Renderer under the hood). 

A quick look at these two options, it seems like `react-native-testing-library` (RNTL) might be easier for us to set up and use. 

So what happens when you do `npm install --save-dev @testing-library/react-native`? 

BOOM! You get an error saying there is a dependency conflict with React, React Native, and React Test Renderer. 

The RNTL team anticipated this, and so they have [a section specifically for this error](https://callstack.github.io/react-native-testing-library/docs/troubleshooting#matching-react-native-react--react-test-renderer-versions) (Note: it was not helpful, as it just points us to GitHub issues with dozens of suggested solutions with mixed of up and down votes. None of those worked for me. 

In hindsight, I should've looked at the [example repository](https://callstack.github.io/react-native-testing-library/docs/troubleshooting#example-repository) earlier and explicitly install `react-test-renderer` to be the same version as react (**this was the key**). 

Ended up needing to: 
1. `npm install --save-dev react-test-renderer@18.1.0` (must be the same version as our react) 
2. `npm install --save-dev @testing-library/react-native` (now it magically works!) 
7. `npm install --save-dev @testing-library/jest-native` ([recommended by RNTL](https://callstack.github.io/react-native-testing-library/docs/getting-started#additional-jest-matchers))
8. Set up `jest.config.ts` according to [this article](https://metareal.blog/en/post/2021/01/16/setup-jest-for-expo-typescript-project/) (we for some reason still have to use `jest-expo` as our jest's preset. I honestly could not find any documentation on this whatsoever and I don't understand why we can't use, say, `react-native` preset) 

## What I learned

The struggle did not go to waste, as I did learn a few things from this: 

1. To find out the exact version of the packages our project have, we can do: 
```zsh
npx envinfo --npmPackages react,react-native,react-test-renderer,@testing-library/react-native
```
Which returns
```zsh
 npmPackages:
    @testing-library/react-native: ^12.0.0 => 12.0.0 
    react: 18.1.0 => 18.1.0 
    react-native: 0.70.5 => 0.70.5 
    react-test-renderer: 18.1.0 => 18.1.0 
```

## How to setup E2E testing with detox locally: 
1. `npm i detox-cli -g`
2. `brew tap wix/brew`
3. `brew install applesimutils`
4. `detox build --configuration ios.sim.debug` (super long)
5. `npm start`
6. If running locally, run `detox test --configuration ios.sim.debug` 
7. If running on GitHub action, run `detox test --configuration ios.sim.release`

### CODE REVIEW

File: __tests__/components/map/UserCard.test.tsx [hunk 1]
```diff
@@ -63,42 +66,42 @@ describe('<UserCard />', () => {
...
+  //     const press2 = screen.getByLabelText(TestingLabel.USER_CARD_CALL_USER);
+  //     const press3 = screen.getByLabelText(TestingLabel.USER_CARD_PING_USER);
+
+  //     // click on the pressables
+  //     fireEvent(press1, 'press');
+  //     fireEvent(press2, 'press');
+  //     fireEvent(press3, 'press');
+
+  //     // TODO: test what happens when each pressables are pressed
+  //   });
```
Comment (jacoblurie29) [inline __tests__/components/map/UserCard.test.tsx:106]: why commented out?

File: jest.setup.js [hunk 2]
```diff
@@ -1,21 +1,38 @@
...
+// // mock registerRootComponent to resolve error
+// jest.mock('expo', () => ({
+//   registerRootComponent: jest.fn(),
+// }));
+
+// // mock useFonts so that it is loaded right away
 // jest.mock('@expo-google-fonts/comfortaa', () => ({
 //   // useFonts will return [true] right away
 //   useFonts: ({}) => [true],
 // }));
```
Comment (jacoblurie29) [inline jest.setup.js:24]: commented out?

File: jest.setup.js [hunk 3]
```diff
@@ -1,21 +1,38 @@
...
+// }));
 
-// jest.mock('react-native/Libraries/EventEmitter/NativeEventEmitter');
+// // mock the react native mapbox
+// jest.mock('@rnmapbox/maps', () => ({
+//   MapView: () => null,
+//   Camera: () => null,
+//   CameraStop: () => null,
+//   setAccessToken: jest.fn(),
+// }));
```
Comment (jacoblurie29) [inline jest.setup.js:38]: see above

File: src/screens/auth/SignInScreen.tsx [hunk 4]
```diff
@@ -95,6 +95,7 @@ const SignInScreen = ({ navigation }: NativeStackScreenProps) => {
         <View style={SignInScreenStyles.inputsContainer}>
           <Text style={SignInScreenStyles.inputsTitle}>Sign in with email</Text>
           <TextInput
+            testID='signin-email-input'
```
Comment (jacoblurie29) [inline src/screens/auth/SignInScreen.tsx:98]: make this an enum or constant

File: src/screens/auth/SignInScreen.tsx [hunk 5]
```diff
@@ -107,6 +108,7 @@ const SignInScreen = ({ navigation }: NativeStackScreenProps) => {
           />
           <View style={SignInScreenStyles.passwordInputContainer}>
             <TextInput
+              testID='signin-password-input'
```
Comment (jacoblurie29) [inline src/screens/auth/SignInScreen.tsx:111]: see above

File: src/screens/auth/SignInScreen.tsx [hunk 6]
```diff
@@ -119,6 +121,7 @@ const SignInScreen = ({ navigation }: NativeStackScreenProps) => {
               onChangeText={setPassword}
             />
             <Pressable
+              testID='signin-password-visibility'
```
Comment (jacoblurie29) [inline src/screens/auth/SignInScreen.tsx:124]: see above

File: src/screens/auth/SignInScreen.tsx [hunk 7]
```diff
@@ -129,6 +132,7 @@ const SignInScreen = ({ navigation }: NativeStackScreenProps) => {
             </Pressable>
           </View>
           <TouchableOpacity
+            testID='signin-forgot-password-button'
```
Comment (jacoblurie29) [inline src/screens/auth/SignInScreen.tsx:135]: above :)))

File: src/screens/auth/SignInScreen.tsx [hunk 8]
```diff
@@ -140,6 +144,7 @@ const SignInScreen = ({ navigation }: NativeStackScreenProps) => {
 
         {/* Sign-In Button */}
         <TouchableOpacity
+          testID='signin-button'
```
Comment (jacoblurie29) [inline src/screens/auth/SignInScreen.tsx:147]: aboveeeee

File: src/screens/auth/SignInScreen.tsx [hunk 9]
```diff
@@ -176,6 +181,7 @@ const SignInScreen = ({ navigation }: NativeStackScreenProps) => {
         <View style={SignInScreenStyles.signUpMessageContainer}>
           <Text style={SignInScreenStyles.signUpPretext}>New here? </Text>
           <TouchableOpacity
+            testID='signin-signup-button'
```
Comment (jacoblurie29) [inline src/screens/auth/SignInScreen.tsx:184]: last one lol

**Review by jacoblurie29 — CHANGES_REQUESTED** [review]

File: __tests__/components/map/UserCard.test.tsx (hunk 1 above)
Comment (zineanteoh) [reply __tests__/components/map/UserCard.test.tsx:106]: these are old tests. need to re-write in the future because our `<App/>` changed significantly since then

File: jest.setup.js (hunk 2 above)
Comment (zineanteoh) [reply jest.setup.js:24]: yeah leaving the code there so that in the future when we need to mock those files we can uncomment it

File: jest.setup.js (hunk 3 above)
Comment (zineanteoh) [reply jest.setup.js:38]: same as above
